### PR TITLE
Make SwiftValue hash/equality work the same as SwiftObject

### PR DESCRIPTION
**Pull request by tbkka**
TL;DR:  Update PR #68720 with lessons learned in reviewing #69464

Background:

* SwiftValue can expose Swift value types (structs/enums) to ObjC by wrapping them in an Obj-C object on the heap

* SwiftObject is the Obj-C type that Swift class objects all inherit from (when viewed from Obj-C).  This allows arbitrary Swift class objects to be passed into Obj-C.

History:

* PR #4124 made Obj-C `-hash` and `-isEqual:` work for SwiftValue for Hashable Swift types

* PR #68720 extended SwiftValue to also support Equatable Swift types

* PR #69464 added similar support to SwiftObject

In the process of working through #69464, we found a better way to handle an ObjC request for `-hash` for a type that is Swift Equatable but not Hashable.  This PR updates SwiftValue to use the same approach.  The approach considers three cases:

1. A Hashable type can forward both `-hash` and `-isEqual:` to the Swift object.

2. A type that is neither Equatable nor Hashable can implement `-isEqual:` as the identity check and `-hash` as returning the address of the object in memory.

3. A type is that Equatable but not Hashable is more complex.

In this last case, we can easily forward `-isEqual:` to the Equatable conformance but ObjC also requires us to always provide a compatible `-hash` implementation. The only way to do this is to have `-hash` return a constant, but that is a very bad idea in general, so we're also including a log message whenever we see a request for `-hash` on a Swift value that is Equatable but not Hashable.
To limit performance problems from the logging itself, we emit the log message only once for each type.

Testing:  I copied the fairly extensive SwiftObjectNSObject test suite over as Swift**Value**NSObject in order to verify the new behaviors and help guard against future regressions.

### CODE REVIEW

**Review by al45tair — APPROVED** [review]
Seems like a good change. I did wonder briefly whether it should be a fatal error to send `-hash` to a Swift object that is `Equatable` but not `Hashable`, but I think on balance the warning approach makes sense.

**Comment by tbkka** [comment]
Yeah, @lorentey also suggested that in the discussion for #69464.  I opted for the warning approach since this isn't a fundamental safety problem.

**Comment by glessard** [comment]
@swift-ci please test

**Comment by tbkka** [comment]
The overly-expansive SwiftValueNSObject test suite had a few cases that aren't really ready.  I commented those out.

**Comment by tbkka** [comment]
@swift-ci Please test

**Comment by tbkka** [comment]
@swift-ci Please test Windows platform

**Comment by tbkka** [comment]
@swift-ci Please test Windows

**Comment by tbkka** [comment]
@swift-ci Please test Windows
